scripts/ubuntu_dev_env_setup.sh

```diff
@@ -1,74 +1,321 @@
 #!/bin/bash
 
+# Exit immediately on error
+set -e
+
+# Colors for messages
+GREEN="\e[32m"
+RED="\e[31m"
+YELLOW="\e[33m"
+RESET="\e[0m"
+
 # Update and upgrade the system
 upgrade_system() {
+    echo -e "${YELLOW}Updating and upgrading the system...${RESET}"
     sudo apt update && sudo apt upgrade -y
 }
 
 # Install essential development tools
 install_dev_tools() {
+    echo -e "${YELLOW}Installing core development tools...${RESET}"
     sudo apt install -y \
         git \
         vim \
         tilix \
         curl \
         wget \
         jq \
-        unzip
+        unzip \
+        gnupg \
+        apt-transport-https \
+        ca-certificates \
+        software-properties-common
 }
 
 # Install VS Code (via Flatpak)
 install_vscode() {
+    echo -e "${YELLOW}Installing Azure CLI...${RESET}"
     flatpak install -y flathub com.visualstudio.code
+    flatpak info com.visualstudio.code || echo -e "${RED}VS Code installation verification failed!${RESET}"
 }
 
-
 # Install Cloud CLIs
 # ====================================
-# Azure 
-
-# AWS
+# Azure CLI
+install_azure_cli() {
+    echo -e "${YELLOW}Installing Azure CLI...${RESET}"
+    curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash
+    # Verify Azure CLI has installed
+    az version || echo -e "${RED}Azure CLI installation verification failed!${RESET}"
+}
 
-# Google
+# AWS CLI
+install_aws_cli() {
+    echo -e "${YELLOW}Installing AWS CLI...${RESET}"
+    curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
+    unzip awscliv2.zip
+    sudo ./aws/install
+    # Verify AWS CLI has installed
+    aws --version || echo -e "${RED}AWS CLI installation verification failed!${RESET}"
+    # Remove installation "junk"
+    rm -rf aws awscliv2.zip
+}
+   
+# Google Cloud SDK
+install_google_cloud_sdk() {
+    echo -e "${YELLOW}Installing Google Cloud SDK...${RESET}"
+    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo gpg --dearmor -o /usr/share/keyrings/cloud.google.gpg
+    echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] https://packages.cloud.google.com/apt cloud-sdk main" | sudo tee -a /etc/apt/sources.list.d/google-cloud-sdk.list
+    sudo apt update && sudo apt install google-cloud-cli
+    # Verify Google Cloud SDK has installed
+    gcloud version || echo -e "${RED}Google Cloud SDK installation verification failed!${RESET}"
+}
 
-#Install Infrastructure as Code tools
+# Install Infrastructure as Code tools
 # ====================================
 # Terraform
+install_terraform() {
+    echo -e "${YELLOW}Installing Terraform...${RESET}"
+
+    wget -O- https://apt.releases.hashicorp.com/gpg | \ 
+    gpg --dearmor | \
+    sudo tee /usr/share/keyrings/hashicorp-archive-keyring.gpg > /dev/null
+
+    echo "deb [signed-by=/usr/share/keyrings/hashicorp-archive-keyring.gpg] \
```

Comment:
ran into issue on this line, I don't think you can split string (within the double quotes) with `\`